hw/dv/dpi/usbdpi/usbdpi.c

```diff
@@ -430,80 +434,76 @@ void readBaud(struct usbdpi_ctx *ctx) {
       ctx->hostSt = HS_SENDACK;
       break;
     case HS_SENDACK:
-      if (ctx->tick_bits == ctx->wait) {
-        ctx->state = ST_SYNC;
-        ctx->bytes = 1;
-        ctx->datastart = -1;
-        ctx->byte = 0;
-        ctx->bit = 1;
-        ctx->data[0] = USB_PID_ACK;
+      if (ctx->tick_bits >= ctx->wait) {
+        transfer_status(ctx, tr, USB_PID_ACK);
         ctx->hostSt = HS_EMPTYDATA;
       }
       break;
     case HS_EMPTYDATA:
-      ctx->state = ST_SYNC;
-      ctx->bytes = 6;
-      ctx->datastart = 3;
-      ctx->byte = 0;
-      ctx->bit = 1;
-      ctx->data[0] = USB_PID_OUT;
-      ctx->data[1] = 0x82;
-      ctx->data[2] = 0 | CRC5(0x82, 11) << 3;
-      ctx->data[3] = USB_PID_DATA1;
-      ctx->data[4] = 0x0;  // pre-computed CRC32
-      ctx->data[5] = 0x0;
+      // Transmit OUT transaction with zero-length DATA packet
+      transfer_token(tr, USB_PID_OUT, ctx->dev_address, ENDPOINT_ZERO);
+      if (INSERT_ERR_DATA_TOGGLE) {
+        // NOTE: This raises a LinkOutErr on the USBDEV because it is expecting
+        //       DATA0
+        // TODO - uint8_t bad_pid = DATA_TOGGLE_ADVANCE(ctx->tx_next_data);
```

Comment:
Does this still occur? It looks like you have changed readBaud() to correctly send the SETUP transaction to endpoint 0 now. Previously, there was a data toggle issue because the SETUP transaction was not ACK'd (because it was sent to endpoint 1, which doesn't answer).

The OUT endpoint _should_ be expecting a DATA1 now. In the RTL, the timing and qualification of the toggle set looks suspicious, though, since I don't see checking whether this is a control endpoint... Might be a bug there :thinking:

Reply:
Oh, never mind. This "read baud" request isn't actually implemented in the software, and the device NAKs the data stage. The response _should_ be a STALL, but that isn't done properly in testutils right now.

Reply:
Do you have a reference for the set/get baud commands please? The only thing I've seen is in the CDC/PSTN specs where SET/GET COMM FEATURE have codes 2 and 3 also, but transposed relative to what the DPI model is using. Also, I've been intrigued by the fact that they are annotated as '_vendor_, endpoint'

I'm not entirely clear what was/is the intent of the readBaud/setBaud of these functions.

Reply:
It's probably better to think of these as not supported and just get rid of them. There is no spec, so far as I'm aware, but if you'd like to see where it comes from...

https://chromium.googlesource.com/chromiumos/platform/ec/+/master/chip/stm32/usb-stream.c#167

The 0x5001 vendor interface maps to a usb-serial-simple driver in Linux, and I believe that driver does not support these control requests.